File: files/jvm/src/main/java/com/swoval/runtime/Platform.java
package com.swoval.runtime;

/** Provides some platform specific properties. */
public class Platform {
  private Platform() {}

  private static final boolean isFreeBSD = System.getProperty("os.name", "").startsWith("FreeBSD");
  private static final boolean isLinuxValue = System.getProperty("os.name", "").startsWith("Linux");
  private static final boolean isMacValue =
      System.getProperty("os.name", "").startsWith("Mac OS X");
  private static final boolean isWinValue = System.getProperty("os.name", "").startsWith("Windows");
  private static final String tmpDirValue = System.getProperty("java.io.tmpdir", "/tmp");

  /**
   * Returns true if the runtime is a java virtual machine.
   *
   * @return true if the runtime is a java virtual machine.
   */
  public static boolean isJVM() {
    return true;
  }

  /**
   * Returns true if running on freebsd.
   *
   * @return true if running on freebsd.
   */
  public static boolean isFreeBSD() {
    return isFreeBSD;
  }
  /**
   * Returns true if running on a mac.
   *
   * @return true if running on a mac.
   */
  public static boolean isMac() {
    return isMacValue;
  }

  /**
   * Returns true if running on windows.
   *
   * @return true if running on windows.
   */
  public static boolean isWin() {
    return isWinValue;
  }

  /**
   * Returns true if running on linux.
   *
   * @return true if running on linux.
   */
  public static boolean isLinux() {
    return isLinuxValue;
  }

  /**
   * Returns the system temporary directory location
   *
   * @return the system temporary directory location
   */
  public static String tmpDir() {
    return tmpDirValue;
  }
}

File: nio-js/src/main/scala/com/swoval/runtime/Platform.scala
package com.swoval.runtime

import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport

/**
 * Provides some platform specific properties.
 */
object Platform {
  private[this] val (_isMac, _isWin, _isLinux, _isFreeBSD) = {
    val platform = os.platform()
    (
      platform == "darwin",
      platform == "win32",
      platform == "linux",
      platform.toLowerCase.startsWith("freebsd")
    )
  }
  def isFreeBSD(): Boolean = false
  def isJVM(): Boolean = false
  def isLinux(): Boolean = _isLinux
  def isMac(): Boolean = _isMac
  def isWin(): Boolean = _isWin
  def tmpDir(): String = os.tmpdir()
}
@js.native
@JSImport("os", JSImport.Default)
private object os extends js.Object {
  def platform(): String = js.native
  def tmpdir(): String = js.native
}
